zrml/global-disputes/src/migrations.rs

```diff
@@ -0,0 +1,398 @@
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+// We use these utilities to prevent having to make the swaps pallet a dependency of
+// prediciton-markets. The calls are based on the implementation of `StorageVersion`, found here:
+// https://github.com/paritytech/substrate/blob/bc7a1e6c19aec92bfa247d8ca68ec63e07061032/frame/support/src/traits/metadata.rs#L168-L230
+// and previous migrations.
+
+extern crate alloc;
+
+use crate::{types::*, Config, Pallet as GDPallet, *};
+use frame_support::{
+    dispatch::Weight,
+    log,
+    pallet_prelude::PhantomData,
+    traits::{Get, OnRuntimeUpgrade, StorageVersion},
+};
+use sp_runtime::traits::Saturating;
+
+#[cfg(feature = "try-runtime")]
+use alloc::string::ToString;
+#[cfg(feature = "try-runtime")]
+use scale_info::prelude::format;
+
+const GD_REQUIRED_STORAGE_VERSION: u16 = 0;
+const GD_NEXT_STORAGE_VERSION: u16 = 1;
+
+pub struct ModifyGlobalDisputesStructures<T>(PhantomData<T>);
+
+impl<T: Config + zrml_market_commons::Config> OnRuntimeUpgrade
+    for ModifyGlobalDisputesStructures<T>
+{
+    fn on_runtime_upgrade() -> Weight
+    where
+        T: Config,
+    {
+        let mut total_weight = T::DbWeight::get().reads(1);
+        let gd_version = StorageVersion::get::<GDPallet<T>>();
+        if gd_version != GD_REQUIRED_STORAGE_VERSION {
+            log::info!(
+                "ModifyGlobalDisputesStructures: global disputes version is {:?}, require {:?};",
+                gd_version,
+                GD_REQUIRED_STORAGE_VERSION,
+            );
+            return total_weight;
+        }
+        log::info!("ModifyGlobalDisputesStructures: Starting...");
+
+        for (market_id, winner_info) in crate::Winners::<T>::drain() {
+            total_weight = total_weight.saturating_add(T::DbWeight::get().reads(1));
+
+            let owners = winner_info.outcome_info.owners;
+            let owners_len = owners.len();
+            let possession = match owners_len {
+                0usize => None,
+                1usize => Some(Possession::Paid {
+                    owner: owners
+                        .get(0)
+                        .expect("Owners len is 1, but could not get this owner.")
+                        .clone(),
+                    fee: T::VotingOutcomeFee::get(),
+                }),
+                _ => Some(Possession::Shared { owners }),
+            };
+
+            let outcome_info =
+                OutcomeInfo { outcome_sum: winner_info.outcome_info.outcome_sum, possession };
+            let gd_info = GDInfo {
+                winner_outcome: winner_info.outcome,
+                outcome_info,
+                status: GDStatus::Finished,
```

Comment:
You tipped your toes on a weak spot. The reason for using Finished here is that I know that the battery station has only finished global disputes. I have a strong conviction that nobody is going to create global dispute until the runtime upgrade happens. This conviction comes from the knowledge that a global dispute requires a lot of technical expertise to get triggered. Even if a global dispute is started in that timespan it's not critical. Why is it not critical? Because there is no real economic value associated with the battery station and if a test is really required, you can just start another global dispute after the update. I know this is not correct, but faster.

Reply:
Yeah, we could clear the entire storage and that would be easier indeed. I should have done this instead. But it is what it is now. I decided to translate the existing storage (although it is non existent, because at the moment the storage is clean). My conviction stays strong that no new global dispute will be initiated (if not you tell someone or you yourself force a global dispute). I see no possible scenario that the chain bricks with the current migration. In the most likely scenario this whole migration has no effect at all, because all storage items affected by this migration are clean anyways. Although I have to admit that this migration was a nice training for me.

Reply:
I'll take another look. I admit that it probably doesn't matter too much here.

Reply:
Yeah, seems fine to me. If worst comes to worst, we'll delete the affected markets.